lib/logstash/inputs/tcp/chooser.rb

```diff
@@ -0,0 +1,72 @@
+# encoding: utf-8
+
+##
+# The `Chooser` provides a safe way to provide multiple distinct sets of behaviours,
+# exactly one of which will be selected at runtime.
+#
+# @api internal
+class LogStash::Inputs::Tcp::Chooser
+  ##
+  # @param supported_choices [Array[Symbol]]
+  def initialize(supported_choices)
+    @supported_choices = supported_choices.dup.freeze
+  end
+
+  ##
+  # @param choice [Symbol]
+  # @param error_name [#to_s] the name of this choice, to be used in case of an error
+  # @param error_class [Exception] the exception class to use in case of an error
+  # @return [Choice]
+  def choose(choice, error_name="choice", error_class=ArgumentError)
+    if !@supported_choices.include?(choice)
+      message = sprintf("unsupported %s `%s`; expected one of %s", error_name, choice.to_s, @supported_choices.map(&:to_s))
+      logger.error(message)
```

Comment:
there seems to be no `logger` method ...

Reply:
I've extracted the "Chooser" to the ECS Support gem in v1.2, and now this PR uses the shared implementation where we have resolved the problem you mention.